vkubelet/sync.go

```diff
@@ -0,0 +1,144 @@
+package vkubelet
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+)
+
+func NewCancellableMutex() *CancellableMutex {
+	return &CancellableMutex{
+		ch: make(chan struct{}, 1),
+	}
+}
+
+// CancellableMutex is a mutex in which the Lock() action can be canncelled by
+// cancelling a context.
+// This must be initialized iby calling `NwwCancellableMutex()`
+//
+// You should expect this to be quite a bit slower than a normal sync.Mutex from
+//n the stdlib.
+type CancellableMutex struct {
+	ch chan struct{}
+}
+
+// Lock locks the mutex.
+// If the passed in context is cancelled before the mutex can be acquired the
+// lock request will be cancelled and an error will be returned.
+//
+// An error is only returned if the context was cancelled.
+func (m *CancellableMutex) Lock(ctx context.Context) error {
+	if m.ch == nil {
+		panic("cancellable mutex is not initialized")
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case m.ch <- struct{}{}:
+	}
+	return nil
+}
+
+// Unlock unlocks the mutex
+// This will panic if the mutex is not locked.
+func (m *CancellableMutex) Unlock() {
+	select {
+	case <-m.ch:
+	default:
+		panic("unlock of unlocked mutex")
+	}
+}
+
+// ErrNoSuchLock is returned when the requested lock does not exist
+var ErrNoSuchLock = errors.New("no such lock")
+
+// NamedLock provides a locking mechanism based on the passed in reference name
+type NamedLock struct {
+	mu    sync.Mutex
```

Comment:
use *sync.Mutex or derive from sync.Mutex

Reply:
This is fine here because we are operating on a a pointer for `NamedLock`, and generally how snyc.Mutex would be initialized in situations like this.